node/summary-calculation-rpc/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+    traits::{Block as BlockT, SaturatedConversion},
+};
+use std::sync::{Arc, Mutex};
+
+use jsonrpsee::{core::RpcResult, proc_macros::rpc};
+
+use node_primitives::AccountId;
+use tnf_service::{
+    extrinsic_utils::{self},
+    merkle_tree_utils::*,
+    summary_utils::EncodedLeafData,
+};
+
+#[rpc(server)]
+pub trait SummaryCalculationProviderRpc {
+    #[method(name = "summary_calculation", blocking)]
+    fn get_summary_calculation(&self, from_block: u32, to_block: u32) -> RpcResult<String>;
+}
+
+const CACHE_PREFIX: &[u8] = b"tnf_summary_cache::v1::";
+
+pub struct SummaryCalculationProvider<C, Block, O = ()> {
+    client: Arc<C>,
+    offchain_storage: Option<Arc<Mutex<O>>>,
+    _marker: std::marker::PhantomData<Block>,
+}
+
+impl<C, Block, O> SummaryCalculationProvider<C, Block, O>
+where
+    O: OffchainStorage,
+{
+    pub fn new(client: Arc<C>, offchain_storage: Option<O>) -> Self {
+        let wrapped_storage = offchain_storage.map(|storage| Arc::new(Mutex::new(storage)));
+        Self { client, offchain_storage: wrapped_storage, _marker: Default::default() }
+    }
+
+    fn get_cached_summary(&self, from_block: u32, to_block: u32) -> Option<[u8; 32]> {
+        if let Some(ref storage_mutex) = self.offchain_storage {
+            if let Ok(storage) = storage_mutex.lock() {
+                let key = (from_block, to_block).encode();
+                storage.get(CACHE_PREFIX, &key).and_then(|data| {
+                    if data.len() == 32 {
+                        let mut array = [0u8; 32];
+                        array.copy_from_slice(&data);
+                        Some(array)
+                    } else {
+                        None
+                    }
+                })
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    
+    fn set_cached_summary(&self, from_block: u32, to_block: u32, root: [u8; 32]) {
+        if let Some(ref storage_mutex) = self.offchain_storage {
+            if let Ok(mut storage) = storage_mutex.lock() {
+                let key = (from_block, to_block).encode();
+                storage.set(CACHE_PREFIX, &key, &root.to_vec());
+            }
+        }
+    }
+
+
+}
+
+impl<C, Block> SummaryCalculationProvider<C, Block, ()> {
+    pub fn new_without_storage(client: Arc<C>) -> Self {
+        Self { client, offchain_storage: None, _marker: Default::default() }
+    }
+}
+
+impl<C, Block, O> SummaryCalculationProviderRpcServer for SummaryCalculationProvider<C, Block, O>
+where
+    Block: BlockT,
+    C: Send + Sync + 'static + BlockBackend<Block> + UsageProvider<Block> + HeaderBackend<Block>,
+    O: OffchainStorage + 'static,
+    AccountId: Clone + std::fmt::Display + Codec,
+{
+    fn get_summary_calculation(&self, from_block: u32, to_block: u32) -> RpcResult<String> {
+        let finalized_block_number: u32 = self.client.info().finalized_number.saturated_into();
+
+        if to_block <= finalized_block_number {
```

Comment:
I would prefer `if to_block > finalized_block_number` was treated as an error, exit early and then treat the cache case.
It simpler to read and more explicit.